src/app/core/json-patch/json-patch-operations.reducer.ts

```diff
@@ -351,7 +351,29 @@ function addOperationToList(body: JsonPatchOperationObject[], actionType, target
       newBody.push(makeOperationEntry({ op: JsonPatchOperationType.move, from: fromPath, path: targetPath }));
       break;
   }
-  return newBody;
+  return dedupeOperationEntries(newBody);
+}
+
+/**
+ * TAMU Customization
+ * Dedupe operation entries by op and path. This prevents processing unnecessary patches in a single PATCH request.
+ *
+ * @param body JSON patch operation object entries
+ * @returns deduped JSON patch operation object entries
+ */
+function dedupeOperationEntries(body: JsonPatchOperationObject[]): JsonPatchOperationObject[] {
+  const ops = new Map<string, any>();
+  for (let i = body.length - 1; i >= 0; i--) {
```

Comment:
This is unsafe.

If `body.length` is `0`, then `0-1=` a really big number or a negative number.

This needs a `body.length == 0` check before the loop.

Reply:
https://github.com/TAMULib/dspace-angular/pull/62/files#diff-463743feea58e1b959f5ff4e4c439124e40a38add3db16de23a08588b13add50R331

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from

Reply:
Here is what Bing says.

<img width="399" alt="image" src="https://github.com/TAMULib/dspace-angular/assets/144841721/c839be50-252b-433c-ab4b-e2677818bd5a">
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/length
https://stackoverflow.com/questions/42708864/what-is-a-safe-length-of-javascript-strings
https://stackoverflow.com/questions/6154989/maximum-size-of-an-array-in-javascript

Reply:
From this:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array
> Normalization of the length property
> 
> The length property is converted to an integer and then clamped to the range between 0 and 253 - 1. NaN becomes 0, so even when length is not present or is undefined, it behaves as if it has value 0.
> 
> The language avoids setting length to an unsafe integer. All built-in methods will throw a TypeError if length will be set to a number greater than 253 - 1. However, because the length property of arrays throws an error if it's set to greater than 232 - 1, the safe integer threshold is usually not reached unless the method is called on a non-array object.

It sounds like the potentially bad case cannot happen based on that.

However, this:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/length

> The value of the length property is a nonnegative integer with a value less than 232.

So it would be unsigned, which would make the `0-1` unsafe.

I think things may be safe based on the first quote above where Javascript might still handle the case where the 0-1 pitfall could result in a large number is prevented.

Furthermore, there is this comment:
> const listC = new Array(-100); // Negative numbers are not allowed

I am going to resolve this change request.

Reply:
@kaladay thanks for the links and details.

In this case, the length of the `body` array is always 1 or greater.

The [newOperation](https://github.com/TAMULib/dspace-angular/blob/tamu-dspace-7_x/src/app/core/json-patch/json-patch-operations.reducer.ts#L210) method is only called for action types starting with [NEW](https://github.com/TAMULib/dspace-angular/blob/tamu-dspace-7_x/src/app/core/json-patch/json-patch-operations.reducer.ts#L77). Within the `newOperation` the body array is [assigned](https://github.com/TAMULib/dspace-angular/blob/tamu-dspace-7_x/src/app/core/json-patch/json-patch-operations.reducer.ts#L212) using a ternary that will always be defined. All conditional paths in [addOperationToList](https://github.com/TAMULib/dspace-angular/blob/tamu-dspace-7_x/src/app/core/json-patch/json-patch-operations.reducer.ts#L330C10-L330C28) will push a value into the body array. Therfore, the array length will always be one or greater and `body.length - 1` is not a concern.